File: lib/domain/agricultural_task/interactors/agricultural_task_list_interactor.rb
# frozen_string_literal: true

module Domain
  module AgriculturalTask
    module Interactors
      class AgriculturalTaskListInteractor < Domain::AgriculturalTask::Ports::AgriculturalTaskListInputPort
        def initialize(output_port:, user_id:, gateway:, user_lookup:)
          @output_port = output_port
          @gateway = gateway
          @user_id = user_id
          @user_lookup = user_lookup
        end

        def call(input_dto = nil)
          input_dto ||= Domain::AgriculturalTask::Dtos::AgriculturalTaskListInput.new(is_admin: false)

          user = @user_lookup.find(@user_id)
          filtered_tasks = list_tasks_for_input(input_dto, user_id: user.id)

          rows = Domain::Shared::Mappers::ReferencableListRowMapper.map_records(user, filtered_tasks)
          @output_port.on_success(rows)
        rescue Domain::Shared::Policies::PolicyPermissionDenied => e
          @output_port.on_failure(e)
        rescue Domain::Shared::Exceptions::RecordNotFound => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end

        private

        def list_tasks_for_input(input_dto, user_id:)
          query = input_dto.query
          unless input_dto.is_admin
            return @gateway.list_user_owned_tasks(user_id: user_id, query: query)
          end

          case input_dto.filter
          when "user"
            @gateway.list_user_owned_tasks(user_id: user_id, query: query)
          when "reference"
            @gateway.list_reference_tasks(query: query)
          else
            @gateway.list_user_and_reference_tasks(user_id: user_id, query: query)
          end
        end
      end
    end
  end
end

File: crates/agrr-domain/src/agricultural_task/interactors/agricultural_task_list_interactor.rs
//! Ruby: `Domain::AgriculturalTask::Interactors::AgriculturalTaskListInteractor`

use crate::agricultural_task::dtos::AgriculturalTaskListInput;
use crate::agricultural_task::entities::AgriculturalTaskEntity;
use crate::agricultural_task::gateways::AgriculturalTaskGateway;
use crate::agricultural_task::ports::{AgriculturalTaskListOutputPort, ListFailure};
use crate::shared::dtos::Error;
use crate::shared::exceptions::{RecordInvalidError, RecordNotFoundError};
use crate::shared::gateways::UserLookupGateway;
use crate::shared::mappers::referencable_list_row_mapper;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

pub struct AgriculturalTaskListInteractor<'a, G, O, L> {
    output_port: &'a mut O,
    gateway: &'a G,
    user_id: i64,
    user_lookup: &'a L,
}

impl<'a, G, O, L> AgriculturalTaskListInteractor<'a, G, O, L>
where
    G: AgriculturalTaskGateway,
    O: AgriculturalTaskListOutputPort,
    L: UserLookupGateway,
{
    pub fn new(output_port: &'a mut O, user_id: i64, gateway: &'a G, user_lookup: &'a L) -> Self {
        Self { output_port, gateway, user_id, user_lookup }
    }

    pub fn call(
        &mut self,
        input: Option<AgriculturalTaskListInput>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let input = input.unwrap_or_else(AgriculturalTaskListInput::default_for_user);
        let user = self.user_lookup.find(self.user_id);
        let query = input.query.as_deref();
        let tasks = match list_tasks_for_input(self.gateway, &input, user.id, query) {
            Ok(t) => t,
            Err(err) => return self.handle_err(err),
        };
        let rows = referencable_list_row_mapper::map_records(&user, tasks);
        self.output_port.on_success(rows);
        Ok(())
    }

    fn handle_err(&mut self, err: Box<dyn std::error::Error + Send + Sync>) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if let Some(policy) = err.downcast_ref::<PolicyPermissionDenied>() {
            self.output_port.on_failure(ListFailure::Policy(*policy));
            return Ok(());
        }
        if err.downcast_ref::<RecordNotFoundError>().is_some() {
            self.output_port
                .on_failure(ListFailure::Error(Error::new("Record not found".to_string())));
            return Ok(());
        }
        match err.downcast::<RecordInvalidError>() {
            Ok(invalid) => {
                self.output_port
                    .on_failure(ListFailure::Error(Error::new(invalid.to_string())));
                Ok(())
            }
            Err(err) => Err(err),
        }
    }
}

fn list_tasks_for_input<G: AgriculturalTaskGateway>(
    gateway: &G,
    input: &AgriculturalTaskListInput,
    user_id: i64,
    query: Option<&str>,
) -> Result<Vec<AgriculturalTaskEntity>, Box<dyn std::error::Error + Send + Sync>> {
    if !input.is_admin {
        return gateway.list_user_owned_tasks(user_id, query);
    }
    match input.filter.as_str() {
        "user" => gateway.list_user_owned_tasks(user_id, query),
        "reference" => gateway.list_reference_tasks(query),
        _ => gateway.list_user_and_reference_tasks(user_id, query),
    }
}
